kernel/univ.ml

```diff
@@ -712,328 +704,496 @@ let rec cleanup_universes g =
-            let node = (repr g u, p) in
-            node :: accu
+      if visited then None
+      else begin
+        visited_strict := UMap.add u.univ strict !visited_strict;
+        try
+          let f typ u' =
+            match traverse (strict && typ = Le) (repr g u') with
+            | None -> ()
+            | Some exp -> raise (Found_explanation ((typ, make u') :: exp))
           in
-          let le_new = List.fold_left fold le_todo arc.le in
-          let () = arc.status <- SetLe in
-          cmp c (arc :: to_revert) lt_todo le_new
-        | u :: lt ->
-          let arc = repr g u in
-          let p = (Lt, make u) :: p in
-          if arc == arcv then
-            if strict then (to_revert, p) else (to_revert, p)
-          else find ((arc, p) :: lt_todo) lt
-        in
-        find [] arc.lt
+          List.iter (f Lt) u.lt;
+          List.iter (f Le) u.le;
+          None
+        with Found_explanation exp -> Some exp
+      end
   in
-  try
-    let (to_revert, c) = cmp [] [] [] [(arcu, [])] in
-    (** Reset all the touched arcs. *)
-    let () = List.iter (fun arc -> arc.status <- Unset) to_revert in
-    List.rev c
-  with e ->
-    (** Unlikely event: fatal error or signal *)
-    let () = cleanup_universes g in
-    raise e
+  let u = repr g u in
+  if u == v then [(Eq, make v.univ)]
+  else match traverse strict u with Some exp -> exp | None -> assert false
 
-let get_explanation strict g arcu arcv =
-  if !Flags.univ_print then Some (get_explanation strict g arcu arcv)
+let get_explanation strict u v g =
+  if !Flags.univ_print then Some (get_explanation strict u v g)
   else None
 
-type fast_order = FastEQ | FastLT | FastLE | FastNLE
-
-let fast_compare_neq strict g arcu arcv =
-  (* [c] characterizes whether arcv has already been related
-     to arcu among the lt_done,le_done universe *)
-  let rec cmp c to_revert lt_todo le_todo = match lt_todo, le_todo with
-  | [],[] -> (to_revert, c)
-  | arc::lt_todo, le_todo ->
-    if arc_is_lt arc then
-      cmp c to_revert lt_todo le_todo
-    else
-      let () = arc.status <- SetLt in
-      process_lt c (arc :: to_revert) lt_todo le_todo arc.lt arc.le
-  | [], arc::le_todo ->
-    if arc == arcv then
-      (* No need to continue inspecting universes above arc:
-	 if arcv is strictly above arc, then we would have a cycle.
-         But we cannot answer LE yet, a stronger constraint may
-	 come later from [le_todo]. *)
-      if strict then cmp FastLE to_revert [] le_todo else (to_revert, FastLE)
-    else
-      if arc_is_le arc then
-        cmp c to_revert [] le_todo
+(* To compare two nodes, we simply do a forward search.
+   We have two ameliorations:
+   - we ignore nodes that ar higher than the destination;
+   - we do a BFS rather than a DFS because we expect to have a short
+       path (typically, the shortest path has length 1)
+ *)
+let search_path strict u v g =
+  (* Purely functional queue with two lists. *)
```

Comment:
Instead of implementing a BFS against an abstract queue interface, I find it actually cleaner to have two separate list arguments, the vertices of the current generation, and the vertices of the next generation. In general this allows to easily insert specific logic at the point of generational change, and also to avoid the `List.rev` call if ordering does not matter (which I think is the case here).